File: ios/Packages/Bolsillo/Sources/BolsilloDomain/UseCases/SuggestCategoryAndAccount.swift
import Foundation

/// Invokes the AI categorization cascade and resolves a best-guess account,
/// falling back gracefully so that saving is never blocked
/// (Constitution Article V — AI must never block saving).
///
/// Confidence cascade:
/// 1. Classifier top result (if any)
/// 2. Last-used category from most recent transaction
/// 3. `nil` (user picks manually)
///
/// Account cascade:
/// 1. Account of most recent transaction
/// 2. First active account from `AccountRepository`
/// 3. Empty string (UI must handle)
public struct SuggestCategoryAndAccount: Sendable {
    private let classifier: any ExpenseClassifier
    private let transactions: any TransactionRepository
    private let accounts: any AccountRepository

    public init(
        classifier: some ExpenseClassifier,
        transactions: some TransactionRepository,
        accounts: some AccountRepository
    ) {
        self.classifier = classifier
        self.transactions = transactions
        self.accounts = accounts
    }

    public func callAsFunction(_ input: ClassificationInput) async -> Suggestion {
        // Fire classifier and last-used fetch concurrently
        async let classResult = classifier.suggest(input)
        async let lastTx = transactions.lastUsed()

        let (result, last) = await (classResult, lastTx)

        // Category cascade
        let categoryId = result.topCategoryId ?? last?.categoryId

        // Account cascade
        let accountId: String
        if let lastId = last?.accountId {
            accountId = lastId
        } else {
            accountId = await firstAccountId()
        }

        return Suggestion(categoryId: categoryId, accountId: accountId, confidence: result.confidence)
    }

    /// Pulls the first emission from the accounts stream and returns the first account's id.
    private func firstAccountId() async -> String {
        for await list in accounts.observeAccounts() {
            return list.first?.id ?? ""
        }
        return ""
    }
}

File: android/domain/bin/main/com/bolsillo/domain/usecase/SuggestCategoryAndAccount.kt
package com.bolsillo.domain.usecase

import com.bolsillo.domain.ai.ClassificationInput
import com.bolsillo.domain.ai.ExpenseClassifier
import com.bolsillo.domain.port.AccountRepository
import com.bolsillo.domain.port.TransactionRepository
import kotlinx.coroutines.flow.firstOrNull
import javax.inject.Inject

/**
 * FR 3,4,17 — pre-fill category + account. Never throws, never blocks save
 * (Article V). Falls back to last-used when the classifier returns nothing.
 */
data class Suggestion(
    val categoryId: String?,
    val accountId: String?,
    val confidence: Double,
)

class SuggestCategoryAndAccount
    @Inject
    constructor(
        private val classifier: ExpenseClassifier,
        private val transactions: TransactionRepository,
        private val accounts: AccountRepository,
    ) {
        suspend operator fun invoke(input: ClassificationInput): Suggestion {
            val classified = runCatching { classifier.suggest(input) }.getOrNull()
            val lastUsed = runCatching { transactions.lastUsed() }.getOrNull()
            val accountId =
                lastUsed?.accountId
                    ?: accounts.observeAccounts().firstOrNull()?.firstOrNull()?.id
            val categoryId =
                classified?.topCategoryId
                    ?: lastUsed?.categoryId
                    ?: "other"
            return Suggestion(
                categoryId = categoryId,
                accountId = accountId,
                confidence = classified?.confidence ?: 0.0,
            )
        }
    }
